### ISSUE TITLE
Chrome-specific PositionalAudio performance issue

### ISSUE THREAD

**Issue by Artyom17** [mrdoob/three.js#15422]
##### Description of the problem
This is a followup to the issue https://github.com/mrdoob/three.js/issues/15344, which was closed for some reasons, but the issue is still there. 

We recently hit the same issue as the original issue: calls to 
```
panner.setPosition( position.x, position.y, position.z );
panner.setOrientation( orientation.x, orientation.y, orientation.z );
```
or to 
```
panner.positionX.linearRampToValueAtTime( position.x, endTime );
panner.positionY.linearRampToValueAtTime( position.y, endTime );
panner.positionZ.linearRampToValueAtTime( position.z, endTime );
panner.orientationX.linearRampToValueAtTime( orientation.x, endTime );
panner.orientationY.linearRampToValueAtTime( orientation.y, endTime );
panner.orientationZ.linearRampToValueAtTime( orientation.z, endTime );
```
becoming slower and slower, killing FPS very quickly WHEN the source is started and stopped constantly.

Strictly speaking, it is NOT a THREE bug but rather the Chrome's one; however, it is easy enough to add a workaround, suggested by @osov: just check if the state 'isPlaying' before calling those functions:
```
THREE.PositionalAudio

updateMatrixWorld: ( function () {
	...
	return function updateMatrixWorld( force ) {
		THREE.Object3D.prototype.updateMatrixWorld.call( this, force );
		if (!this.isPlaying)        // this is 
			return;                 // the fix
		var panner = this.panner;
	...
	};
} )()
```

We've also filed a bugreport to Chromium Bugs Database: https://bugs.chromium.org/p/chromium/issues/detail?id=914608

There is a repro case for the issue in pure-JS attached to that google issue: https://bugs.chromium.org/p/chromium/issues/attachmentText?aid=371486

##### Three.js version

- [x] Dev
- [x] r99
- [x] r97

##### Browser

- [ ] All of them
- [x ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

**Comment by mrdoob** [comment]
The fix looks good to me. Would you like to do a PR with it? Happy to do it myself otherwise.

**Comment by Artyom17** [comment]
> The fix looks good to me. Would you like to do a PR with it? Happy to do it myself otherwise.

https://github.com/mrdoob/three.js/pull/15424

**Comment by hoch** [comment]
I have not investigated, but the reporter confirmed that the problem is gone in Chrome beta:
https://bugs.chromium.org/p/chromium/issues/detail?id=914608#c6

**Comment by Mugen87** [comment]
> The context must be activated upon user input.

`three.js` creates its internal audio context lazily when `AudioContext.getContext()` is executed for the first time. This usually happens when users create an instance of `AudioListener`, `Audio` or `PositionalAudio`. As long as this object creation is performed after a user interaction, things should work fine, right?

In our demos, we are doing the initialization after a button click. For example here:

https://threejs.org/examples/webaudio_timing.html

Can you confirm that this code demonstrates a valid approach for creating the audio context and using the Web Audio API?

**Comment by Mugen87** [comment]
Sry, let me provide you some two code snippets:

As you can see here, when the button is clicked we execute our initialization routine of the example.

https://github.com/mrdoob/three.js/blob/f55b3d3b91b0f50c28523ed707baec91ddff465a/examples/webaudio_timing.html#L100-L101

In this `init()` function we have the following code which creates the audio listener object which is usually  a child of the camera.

https://github.com/mrdoob/three.js/blob/f55b3d3b91b0f50c28523ed707baec91ddff465a/examples/webaudio_timing.html#L143-L144

The ctor of `AudioListener` then calls `AudioContext.getContext()` in order to have a reference to the raw audio context. 

https://github.com/mrdoob/three.js/blob/f55b3d3b91b0f50c28523ed707baec91ddff465a/src/audio/AudioListener.js#L11-L17

Since it's the first call of `AudioContext.getContext()`  in this example, the audio context is created at that point.

https://github.com/mrdoob/three.js/blob/f55b3d3b91b0f50c28523ed707baec91ddff465a/src/audio/AudioContext.js#L7-L19

> I assume this audio object is some sorts of abstraction on top of AudioBufferSourceNode. 

Correct. These objects always have to be created with the central audio listener of the app. In this way, the get access to the raw web audio context.

**Comment by hoch** [comment]
You can try two things to see if the audio works correctly:

1. Try to print out `context.state` after `AudioContext.getContext()`. You should see `running` if everything works out. If not, somehow user activation is not being delivered to the context's constructor. (and that might be a chrome bug)
2. In Chrome, Go to Setting > Content Settings > Sound > Allow, try to add your page's URL. If the option 1 doesn't work whereas this forced setting works, then it confirms that this might be a bug in Chrome.

**Comment by hoch** [comment]
Great! (otherwise it would've been another bug for me...)

But the question remains; are you still experiencing the performance hit?

**Comment by Mugen87** [comment]
@Artyom17 said

> calls to .... becoming slower and slower, killing FPS very quickly WHEN the source is started and stopped constantly.

I have to test the start/stop case. The mentioned [demo](https://threejs.org/examples/webaudio_timing.html) (which does not perform a start/stop) keeps a good performance over time.

**Comment by Mugen87** [comment]
I've created the following live demo for testing: https://jsfiddle.net/kqdcom1a/1/

I was not able to see any form of performance degradation on Chromium 71.0.3578.80 on Ubuntu and `three.js` `R99`. I've also seen nothing special in the mentioned tracing tool, too. Well, I'm not familiar with it so I've might overseen something.

In any event, I did not seen something unusual in the browser's internal task manager.

BTW: When opening `chrome://tracing `, I'm not able to select `webaudio` in the record categories. The respective checkbox does not exist.

**Comment by hoch** [comment]
Just tested with Canary 73.0.3642.0 (MacOS). I don't see any anomaly or back pressure. Everything should run smoothly.

![screen shot 2018-12-17 at 1 54 40 pm](https://user-images.githubusercontent.com/676891/50118041-55d9b800-0203-11e9-89f5-757efa00baf3.png)

Without the `webaudio` category, tracing the code is meaningless. Not sure why that doesn't show up for you.

**Comment by Artyom17** [comment]
The issue seems to be fixed since Chrome 72.0.3626.10. However, there are a lot of Chrome and Chromium-based browsers below this version, therefore the workaround that was landed is still necessary.

**Comment by Mugen87** [comment]
The PR is definitely a nice addition since it improves performance when positional audios are not playing.
